src/controls/ArrayControl.tsx

```diff
@@ -28,7 +35,7 @@ export function ObjectArrayControl({
   rootSchema,
   uischemas,
   required,
-}: ArrayLayoutProps) {
+}: ArrayLayoutProps | ArrayControlProps) {
```

Comment:
This is a smell to me--have you tried splitting this into two components that wrap ancomponent with common logic? Ie create the datasource array in the higher-level components and pass to the shared component?